File: hoomd/variant.py
# Copyright (c) 2009-2022 The Regents of the University of Michigan.
# Part of HOOMD-blue, released under the BSD 3-Clause License.

"""Define quantities that vary over the simulation.

A `Variant` object represents a scalar function of the time step. Some
operations accept `Variant` values for certain parameters, such as the
``kT`` parameter to `hoomd.md.methods.NVT`.

Use one of the built in variant types, or define your own custom function
in Python:

.. code:: python

    class CustomVariant(hoomd.variant.Variant):
        def __init__(self):
            hoomd.variant.Variant.__init__(self)

        def __call__(self, timestep):
            return (float(timestep)**(1 / 2))

        def _min(self):
            return 0.0

        def _max(self):
            return float('inf')

Note:
    Provide the minimum and maximum values in the ``_min`` and ``_max``
    methods respectively.
"""

from hoomd import _hoomd


class Variant(_hoomd.Variant):
    """Variant base class.

    Variants are scalar valued functions of the simulation time step.

    .. py:method:: __call__(timestep)

        Evaluate the function.

        :param timestep: The time step.
        :type timestep: int
        :return: The value of the function at the given time step.
        :rtype: float
    """

    @property
    def min(self):
        """The minimum value of this variant for :math:`t \\in [0,\\infty)`."""
        return self._min()

    @property
    def max(self):
        """The maximum value of this variant for :math:`t \\in [0,\\infty)`."""
        return self._max()

    def __getstate__(self):
        """Get the variant's ``__dict__`` attribute."""
        return self.__dict__

    def __setstate__(self, state):
        """Restore the state of the variant."""
        _hoomd.Variant.__init__(self)
        self.__dict__ = state

    def _private_eq(self, other):
        """Return whether two variants are equivalent."""
        if not isinstance(other, Variant):
            return NotImplemented
        if not isinstance(other, type(self)):
            return False
        return all(
            getattr(self, attr) == getattr(other, attr)
            for attr in self._eq_attrs)


class Constant(_hoomd.VariantConstant, Variant):
    """A constant value.

    Args:
        value (float): The value.

    `Constant` returns `value` at all time steps.

    Attributes:
        value (float): The value.
    """
    _eq_attrs = ("value",)

    def __init__(self, value):
        Variant.__init__(self)
        _hoomd.VariantConstant.__init__(self, value)

    __eq__ = Variant._private_eq


class Ramp(_hoomd.VariantRamp, Variant):
    """A linear ramp.

    Args:
        A (float): The start value.
        B (float): The end value.
        t_start (int): The start time step.
        t_ramp (int): The length of the ramp.

    `Ramp` holds the value *A* until time *t_start*. Then it ramps linearly from
    *A* to *B* over *t_ramp* steps and holds the value *B*.

    .. image:: variant-ramp.svg
       :alt: Example plot of a ramp variant.

    Attributes:
        A (float): The start value.
        B (float): The end value.
        t_start (int): The start time step.
        t_ramp (int): The length of the ramp.
    """
    _eq_attrs = ("A", "B", "t_start", "t_ramp")

    def __init__(self, A, B, t_start, t_ramp):
        Variant.__init__(self)
        _hoomd.VariantRamp.__init__(self, A, B, t_start, t_ramp)

    __eq__ = Variant._private_eq


class Cycle(_hoomd.VariantCycle, Variant):
    """A cycle of linear ramps.

    Args:
        A (float): The first value.
        B (float): The second value.
        t_start (int): The start time step.
        t_A (int): The hold time at the first value.
        t_AB (int): The time spent ramping from A to B.
        t_B (int): The hold time at the second value.
        t_BA (int): The time spent ramping from B to A.

    `Cycle` holds the value *A* until time *t_start*. It continues holding that
    value until *t_start + t_A*. Then it ramps linearly from *A* to *B* over
    *t_AB* steps and holds the value *B* for *t_B* steps. After this, it ramps
    back from *B* to *A* over *t_BA* steps and repeats the cycle starting with
    *t_A*. `Cycle` repeats this cycle indefinitely.

    .. image:: variant-cycle.svg
       :alt: Example plot of a cycle variant.

    Attributes:
        A (float): The first value.
        B (float): The second value.
        t_start (int): The start time step.
        t_A (int): The holding time at A.
        t_AB (int): The time spent ramping from A to B.
        t_B (int): The holding time at B.
        t_BA (int): The time spent ramping from B to A.
    """
    _eq_attrs = ("A", "B", "t_start", "t_A", "t_AB", "t_B", "t_BA")

    def __init__(self, A, B, t_start, t_A, t_AB, t_B, t_BA):
        Variant.__init__(self)
        _hoomd.VariantCycle.__init__(self, A, B, t_start, t_A, t_AB, t_B, t_BA)

    __eq__ = Variant._private_eq


class Power(_hoomd.VariantPower, Variant):
    """An approach from initial to final value following ``t**power``.

    Args:
        A (float): The start value.
        B (float): The end value.
        power (float): The power of the approach to ``B``.
        t_start (int): The start time step.
        t_ramp (int): The length of the ramp.

    `Power` holds the value *A* until time *t_start*. Then it progresses at
    :math:`t^{\\mathrm{power}}` from *A* to *B* over *t_ramp* steps and holds
    the value *B* after that.

    .. code-block:: python

        p = Power(A=2, B-8, power=1 / 10, t_start=10, t_ramp=20)

    .. image:: variant-power.svg
       :alt: Example plot of a power variant.

    Attributes:
        A (float): The start value.
        B (float): The end value.
        power (float): The power of the approach to ``B``.
        t_start (int): The start time step.
        t_ramp (int): The length of the ramp.
    """
    _eq_attrs = ("A", "B", "power", "t_start", "t_ramp")

    def __init__(self, A, B, power, t_start, t_ramp):
        Variant.__init__(self)
        _hoomd.VariantPower.__init__(self, A, B, power, t_start, t_ramp)

    __eq__ = Variant._private_eq

File: hoomd/Variant.cc
// Copyright (c) 2009-2024 The Regents of the University of Michigan.
// Part of HOOMD-blue, released under the BSD 3-Clause License.

#include "Variant.h"

namespace hoomd
    {
//* Trampoline for classes inherited in python
class VariantPy : public Variant
    {
    public:
    // Inherit the constructors
    using Variant::Variant;

    // trampoline method
    Scalar operator()(uint64_t timestep) override
        {
        PYBIND11_OVERLOAD_NAME(Scalar,     // Return type
                               Variant,    // Parent class
                               "__call__", // name of function in python
                               operator(), // Name of function in C++
                               timestep    // Argument(s)
        );
        }

    Scalar min() override
        {
        PYBIND11_OVERLOAD_PURE_NAME(Scalar,  // Return type
                                    Variant, // Parent class
                                    "_min",  // name of function in python
                                    min      // name of function
        );
        }

    Scalar max() override
        {
        PYBIND11_OVERLOAD_PURE_NAME(Scalar,  // Return type
                                    Variant, // Parent class
                                    "_max",  // name of function in python
                                    max      // name of function
        );
        }
    };

namespace detail
    {
// These testVariant{Method} functions allow us to test that Python custom
// variants work properly in C++. This ensures we can test that the function
// itself can be called in C++ when defined in Python.

/// Method to enable unit testing of C++ variant calls from pytest
Scalar testVariantCall(std::shared_ptr<Variant> t, uint64_t step)
    {
    return (*t)(step);
    }

/// Method to enable unit testing of C++ variant min class from pytest
Scalar testVariantMin(std::shared_ptr<Variant> t)
    {
    return t->min();
    }

/// Method to enable unit testing of C++ variant max class from pytest
Scalar testVariantMax(std::shared_ptr<Variant> t)
    {
    return t->max();
    }

void export_Variant(pybind11::module& m)
    {
    pybind11::class_<Variant, VariantPy, std::shared_ptr<Variant>>(m, "Variant")
        .def(pybind11::init<>())
        .def("__call__", &Variant::operator())
        .def("_min", &Variant::min)
        .def("_max", &Variant::max)
        .def_property_readonly("range", &Variant::range);

    pybind11::class_<VariantConstant, Variant, std::shared_ptr<VariantConstant>>(m,
                                                                                 "VariantConstant")
        .def(pybind11::init<Scalar>(), pybind11::arg("value"))
        .def_property("value", &VariantConstant::getValue, &VariantConstant::setValue)
        .def(pybind11::pickle(
            [](const VariantConstant& variant) { return pybind11::make_tuple(variant.getValue()); },
            [](pybind11::tuple params) { return VariantConstant(params[0].cast<Scalar>()); }));

    pybind11::class_<VariantRamp, Variant, std::shared_ptr<VariantRamp>>(m, "VariantRamp")
        .def(pybind11::init<Scalar, Scalar, uint64_t, uint64_t>(),
             pybind11::arg("A"),
             pybind11::arg("B"),
             pybind11::arg("t_start"),
             pybind11::arg("t_ramp"))
        .def_property("A", &VariantRamp::getA, &VariantRamp::setA)
        .def_property("B", &VariantRamp::getB, &VariantRamp::setB)
        .def_property("t_start", &VariantRamp::getTStart, &VariantRamp::setTStart)
        .def_property("t_ramp", &VariantRamp::getTRamp, &VariantRamp::setTRamp)
        .def(pybind11::pickle(
            [](const VariantRamp& variant)
            {
                return pybind11::make_tuple(variant.getA(),
                                            variant.getB(),
                                            variant.getTStart(),
                                            variant.getTRamp());
            },
            [](pybind11::tuple params)
            {
                return VariantRamp(params[0].cast<Scalar>(),
                                   params[1].cast<Scalar>(),
                                   params[2].cast<uint64_t>(),
                                   params[3].cast<uint64_t>());
            }));

    pybind11::class_<VariantCycle, Variant, std::shared_ptr<VariantCycle>>(m, "VariantCycle")
        .def(pybind11::init<Scalar, Scalar, uint64_t, uint64_t, uint64_t, uint64_t, uint64_t>(),
             pybind11::arg("A"),
             pybind11::arg("B"),
             pybind11::arg("t_start"),
             pybind11::arg("t_A"),
             pybind11::arg("t_AB"),
             pybind11::arg("t_B"),
             pybind11::arg("t_BA"))

        .def_property("A", &VariantCycle::getA, &VariantCycle::setA)
        .def_property("B", &VariantCycle::getB, &VariantCycle::setB)
        .def_property("t_start", &VariantCycle::getTStart, &VariantCycle::setTStart)
        .def_property("t_A", &VariantCycle::getTA, &VariantCycle::setTA)
        .def_property("t_AB", &VariantCycle::getTAB, &VariantCycle::setTAB)
        .def_property("t_B", &VariantCycle::getTB, &VariantCycle::setTB)
        .def_property("t_BA", &VariantCycle::getTBA, &VariantCycle::setTBA)
        .def(pybind11::pickle(
            [](const VariantCycle& variant)
            {
                return pybind11::make_tuple(variant.getA(),
                                            variant.getB(),
                                            variant.getTStart(),
                                            variant.getTA(),
                                            variant.getTAB(),
                                            variant.getTB(),
                                            variant.getTBA());
            },
            [](pybind11::tuple params)
            {
                return VariantCycle(params[0].cast<Scalar>(),
                                    params[1].cast<Scalar>(),
                                    params[2].cast<uint64_t>(),
                                    params[3].cast<uint64_t>(),
                                    params[4].cast<uint64_t>(),
                                    params[5].cast<uint64_t>(),
                                    params[6].cast<uint64_t>());
            }));

    pybind11::class_<VariantPower, Variant, std::shared_ptr<VariantPower>>(m, "VariantPower")
        .def(pybind11::init<Scalar, Scalar, double, uint64_t, uint64_t>(),
             pybind11::arg("A"),
             pybind11::arg("B"),
             pybind11::arg("power"),
             pybind11::arg("t_start"),
             pybind11::arg("t_ramp"))
        .def_property("A", &VariantPower::getA, &VariantPower::setA)
        .def_property("B", &VariantPower::getB, &VariantPower::setB)
        .def_property("power", &VariantPower::getPower, &VariantPower::setPower)
        .def_property("t_start", &VariantPower::getTStart, &VariantPower::setTStart)
        .def_property("t_ramp", &VariantPower::getTRamp, &VariantPower::setTRamp)
        .def(pybind11::pickle(
            [](const VariantPower& variant)
            {
                return pybind11::make_tuple(variant.getA(),
                                            variant.getB(),
                                            variant.getPower(),
                                            variant.getTStart(),
                                            variant.getTRamp());
            },
            [](pybind11::tuple params)
            {
                return VariantPower(params[0].cast<Scalar>(),
                                    params[1].cast<Scalar>(),
                                    params[2].cast<double>(),
                                    params[3].cast<uint64_t>(),
                                    params[4].cast<uint64_t>());
            }));

    m.def("_test_variant_call", &testVariantCall);
    m.def("_test_variant_min", &testVariantMin);
    m.def("_test_variant_max", &testVariantMax);
    }

    } // end namespace detail

    } // end namespace hoomd
